onnxruntime/test/shared_lib/test_inference.cc

```diff
@@ -648,6 +651,454 @@ TEST(CApiTest, multiple_varied_input_custom_op_handler) {
+    ASSERT_EQ(type_info.GetElementCount(), 1u);
+    ASSERT_EQ(type_info.GetElementType(), ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT);
+
+    const float* out_ptr = ort_output.GetTensorData<float>();
+    ASSERT_EQ(out_ptr[0], float_inps[0]);
+  }
+
+  // Second output should be a int64_t.
+  {
+    auto& ort_output = ort_outputs[1];
+    auto type_info = ort_output.GetTensorTypeAndShapeInfo();
+    ASSERT_EQ(type_info.GetShape(), input_dims);
+    ASSERT_EQ(type_info.GetElementCount(), 1u);
+    ASSERT_EQ(type_info.GetElementType(), ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_INT64);
+
+    const int64_t* out_ptr = ort_output.GetTensorData<int64_t>();
+    ASSERT_EQ(out_ptr[0], int_inps[0]);
+  }
+
+  // Last output should be a string.
+  {
+    auto& ort_output = ort_outputs[2];
+    auto type_info = ort_output.GetTensorTypeAndShapeInfo();
+    ASSERT_EQ(type_info.GetShape(), input_dims);
+    ASSERT_EQ(type_info.GetElementCount(), 1u);
+    ASSERT_EQ(type_info.GetElementType(), ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_STRING);
+
+    const size_t str_len = ort_output.GetStringTensorElementLength(0);
+    ASSERT_EQ(str_len, str_input.size());
+
+    std::string str;
+    str.resize(str_len);
+
+    ort_output.GetStringTensorElement(str_len, 0, str.data());
+    ASSERT_EQ(str, str_input);
+  }
+}
+
+TEST(CApiTest, invalid_variadic_input_not_last_custom_op) {
+  // Create an invalid custom op with 2 inputs. The first input is variadic and the last is not.
+  // Expect an error because only the last input may be marked as variadic.
+  TemplatedCustomOp<MyCustomStringLengthsKernel> custom_op(
+      "VariadicNode",
+      // Input config
+      {ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_STRING,
+       ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_STRING},
+      {OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_VARIADIC,
+       OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_REQUIRED},
+      1,
+      true,
+      // Output config
+      {ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_INT64,
+       ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_INT64},
+      {OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_REQUIRED,
+       OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_VARIADIC},
+      1,
+      true);
+
+  Ort::CustomOpDomain custom_op_domain("test");
+  custom_op_domain.Add(&custom_op);
+
+  Ort::SessionOptions session_options;
+  session_options.Add(custom_op_domain);
+
+  bool caught_exception = false;
+
+  try {
+    Ort::Session session(*ort_env, VARIADIC_INPUT_OUTPUT_CUSTOM_OP_MODEL_URI, session_options);
+  } catch (const Ort::Exception& excpt) {
+    caught_exception = true;
+    std::string_view exception_msg = excpt.what();
+    std::string_view expected_err = "Only the last input to a custom op may be marked variadic.";
+    ASSERT_TRUE(exception_msg.find(expected_err) != std::string_view::npos);
+  }
```

Comment:
nit: could use [HasSubstr](https://github.com/google/googletest/blob/main/docs/reference/matchers.md#string-matchers) for the string check and FAIL() after the session ctor instead of `caught_exception`